crypto/ec/ecx_key.c

```diff
@@ -72,11 +72,13 @@ void ossl_ecx_key_free(ECX_KEY *key)
     if (key == NULL)
         return;
 
-    CRYPTO_DOWN_REF(&key->references, &i, key->lock);
-    REF_PRINT_COUNT("ECX_KEY", key);
-    if (i > 0)
-        return;
-    REF_ASSERT_ISNT(i < 0);
+    if (!key->genbyengine) {
```

Comment:
This looks a bit hacky. I don't see any evidence of this kind of engine change for other key types such as RSA and EC keys, so I am not sure why we would accept it here.

Reply:
Just like what I explained in #20505, in OpenSSL 1.1.1, the EVP_PKEY_free() will only decrease the reference of the ECX_KEY one time at the beginning of entry point. Because the amethod isn't null, so it will finally invoke the ecx_free() to free the pkey. But in OpenSSL 3, the reference of the pkey not only decreased at the entry point of EVP_PKEY_free(), but also decreased again in ecx_free(), which invokes the ossl_ecx_key_free(). That's why I changed the code here. 

I don't know why the ecx_free() should be changed like this in OpenSSL 3, if you have any idea to solve the issue #20505, please let me know, thanks!

Reply:
This will only introduce a leak of the `ECX_KEY` since `EVP_PKEY_free` frees the `EVP_PKEY` now, but the reference to the `ECX_KEY` is lost forever.  The real fix is not inside OpenSSL, but inside the engine.